File: python/src/nodes/core/structural_index.py
from __future__ import annotations

import math
from collections.abc import Iterable
from copy import deepcopy
from dataclasses import dataclass
from typing import Literal

from pydantic import ValidationError as PydanticValidationError

from nodes.core.errors import CollisionError, IdError
from nodes.core.ids import NodeId
from nodes.core.node import Node
from nodes.core.relations import Relation
from nodes.core.shapes import EDGES, KEYS, MEMBERSHIP, ORDER

Role = Literal[
    "relation_source",
    "relation_target",
    "membership_member",
    "edges_source",
    "edges_target",
    "order_member",
    "keys_value",
]

_STRUCTURAL_ENTRY_KEYS = frozenset({"uid", "id", "kind", "deprecated_ids", "relations", "structural_refs"})
_STRUCTURAL_REF_ROLES = frozenset({"membership_member", "edges_source", "edges_target", "order_member", "keys_value"})


@dataclass
class OutRef:
    ref: str
    role: Role
    relation: Relation | None = None  # present iff role startswith "relation_"


@dataclass
class InRef:
    source_uid: str
    out_ref: OutRef


@dataclass
class IndexEntry:
    uid: str
    id: str
    kind: str
    deprecated_ids: frozenset[str]
    out_refs: list[OutRef]


@dataclass
class ResolvedEdge:
    relation: Relation
    source_uid: str | None
    target_uid: str | None


def _relation_out_refs(relations: list[Relation]) -> list[OutRef]:
    refs: list[OutRef] = []
    for rel in relations:
        refs.append(OutRef(ref=rel.source, role="relation_source", relation=rel))
        refs.append(OutRef(ref=rel.target, role="relation_target", relation=rel))
    return refs


def _structural_out_refs(node: Node) -> list[OutRef]:
    """Refs from the built-in structural facets. Read directly from `node.facets`
    (registry-independent); they populate `in_refs` for rename + dangling integrity but
    are never relation-graph edges (their `relation` is None)."""
    refs: list[OutRef] = []
    mem = node.facets.get(MEMBERSHIP)
    if isinstance(mem, dict) and isinstance(mem.get("members"), list):
        for m in mem["members"]:
            if isinstance(m, str):
                refs.append(OutRef(ref=m, role="membership_member"))
    eg = node.facets.get(EDGES)
    if isinstance(eg, dict):
        for edge in eg.get("edges", []) or []:
            if isinstance(edge, dict):
                if isinstance(edge.get("source"), str):
                    refs.append(OutRef(ref=edge["source"], role="edges_source"))
                if isinstance(edge.get("target"), str):
                    refs.append(OutRef(ref=edge["target"], role="edges_target"))
    od = node.facets.get(ORDER)
    if isinstance(od, dict):
        for m in od.get("order", []) or []:
            if isinstance(m, str):
                refs.append(OutRef(ref=m, role="order_member"))
    ky = node.facets.get(KEYS)
    if isinstance(ky, dict):
        keys = ky.get("keys", {})
        if isinstance(keys, dict):
            for v in keys.values():
                if isinstance(v, str):
                    refs.append(OutRef(ref=v, role="keys_value"))
    return refs


def _extract_out_refs(node: Node) -> list[OutRef]:
    return _relation_out_refs(node.relations) + _structural_out_refs(node)


def _validated_deprecated_ids(raw: object, entry_id: str) -> list[str]:
    if not isinstance(raw, list):
        raise ValueError("structural snapshot: deprecated_ids must be a list of strings")

    deprecated_ids: list[str] = []
    seen: set[str] = set()
    for dep in raw:
        if not isinstance(dep, str):
            raise ValueError("structural snapshot: deprecated_ids must be a list of strings")
        if dep == entry_id:
            raise ValueError(
                f"structural snapshot: identity claim {dep!r} is both live and deprecated in one entry"
            )
        if dep in seen:
            raise ValueError(
                f"structural snapshot: duplicate deprecated identity claim {dep!r} in one entry"
            )
        seen.add(dep)
        deprecated_ids.append(dep)
    return deprecated_ids


def _validated_structural_refs(raw: object) -> list[OutRef]:
    if not isinstance(raw, list):
        raise ValueError("structural snapshot: structural_refs must be a list")
    out: list[OutRef] = []
    for item in raw:
        if not isinstance(item, dict):
            raise ValueError("structural snapshot: structural_ref must be a dict")
        ref = item.get("ref")
        role = item.get("role")
        if not isinstance(ref, str):
            raise ValueError("structural snapshot: structural_ref ref must be a string")
        if role not in _STRUCTURAL_REF_ROLES:
            raise ValueError("structural snapshot: structural_ref role is invalid")
        out.append(OutRef(ref=ref, role=role))  # type: ignore[arg-type]
    return out


def _validate_snapshot_weight(raw: dict, label: str) -> None:
    if "weight" not in raw:
        return
    weight = raw["weight"]
    if (
        isinstance(weight, bool)
        or not isinstance(weight, (int, float))
        or not math.isfinite(weight)
    ):
        raise ValueError(f"structural snapshot: {label} weight must be a finite number")


def _validate_snapshot_directed(raw: dict, label: str) -> None:
    if "directed" in raw and not isinstance(raw["directed"], bool):
        raise ValueError(f"structural snapshot: {label} directed must be a bool")


class Index:
    """In-memory structural index. Pure data; no file I/O."""

    def __init__(self) -> None:
        self.by_uid: dict[str, IndexEntry] = {}
        self.id_to_uid: dict[str, str] = {}
        self.deprecated_to_uid: dict[str, str] = {}
        self.in_refs: dict[str, list[InRef]] = {}

    @classmethod
    def build(cls, nodes: Iterable[Node]) -> "Index":
        idx = cls()
        for node in nodes:
            if node.uid in idx.by_uid:
                raise CollisionError(f"duplicate uid {node.uid!r} in corpus")
            idx.assert_addable(node)  # fail-early on a corrupt corpus (collision contract)
            idx.upsert(node)
        return idx

    def resolve_uid(self, ref: str) -> str | None:
        return self.id_to_uid.get(ref) or self.deprecated_to_uid.get(ref)

    def assert_addable(self, node: Node) -> None:
        existing = self.by_uid.get(node.uid)
        if existing is not None and existing.id != node.id:
            raise CollisionError(
                f"uid {node.uid!r} already belongs to live id {existing.id!r}; use rename()"
            )
        for claim in (node.id, *node.deprecated_ids):
            owner = self.resolve_uid(claim)
            if owner is not None and owner != node.uid:
                raise CollisionError(f"identity claim {claim!r} already in use by uid {owner!r}")

    def upsert(self, node: Node) -> None:
        if node.uid in self.by_uid:
            self._drop(node.uid)
        entry = IndexEntry(
            uid=node.uid,
            id=node.id,
            kind=node.kind,
            deprecated_ids=frozenset(node.deprecated_ids),
            out_refs=_extract_out_refs(node),
        )
        self.by_uid[node.uid] = entry
        self.id_to_uid[node.id] = node.uid
        for dep in node.deprecated_ids:
            self.deprecated_to_uid[dep] = node.uid
        for oref in entry.out_refs:
            self.in_refs.setdefault(oref.ref, []).append(InRef(source_uid=node.uid, out_ref=oref))

    def to_dict(self) -> dict:
        entries = []
        for entry in self.by_uid.values():
            relations = []
            for o in entry.out_refs:
                if o.role != "relation_source" or o.relation is None:
                    continue
                relation = {
                    "source": o.relation.source,
                    "predicate": o.relation.predicate,
                    "target": o.relation.target,
                    "directed": o.relation.directed,
                    "attrs": deepcopy(o.relation.attrs),
                }
                if o.relation.weight is not None:
                    relation["weight"] = o.relation.weight
                relations.append(relation)
            entries.append(
                {
                    "uid": entry.uid,
                    "id": entry.id,
                    "kind": entry.kind,
                    "deprecated_ids": sorted(entry.deprecated_ids),
                    "relations": relations,
                    "structural_refs": [
                        {"ref": o.ref, "role": o.role}
                        for o in entry.out_refs
                        if not o.role.startswith("relation_")
                    ],
                }
            )
        return {"entries": entries}

    @classmethod
    def from_dict(cls, d: dict) -> "Index":
        if not isinstance(d, dict):
            raise ValueError("structural snapshot: document must be a dict")
        if "entries" not in d:
            raise ValueError("structural snapshot: missing entries")
        entries_raw = d["entries"]
        if not isinstance(entries_raw, list):
            raise ValueError("structural snapshot: entries must be a list")
        idx = cls()
        for raw in entries_raw:
            if not isinstance(raw, dict):
                raise ValueError("structural snapshot: entry must be a dict")
            missing = _STRUCTURAL_ENTRY_KEYS - raw.keys()
            if missing:
                raise ValueError(f"structural snapshot: entry missing {sorted(missing)[0]}")
            uid = raw["uid"]
            entry_id = raw["id"]
            kind = raw["kind"]
            relations_raw = raw["relations"]
            structural_refs_raw = raw["structural_refs"]
            if not isinstance(uid, str):
                raise ValueError("structural snapshot: entry uid must be a string")
            if not isinstance(entry_id, str):
                raise ValueError("structural snapshot: entry id must be a string")
            if not isinstance(kind, str):
                raise ValueError("structural snapshot: entry kind must be a string")
            try:
                parsed_id = NodeId.parse(entry_id)
            except IdError as exc:
                raise ValueError("structural snapshot: entry id must be a valid node id") from exc
            if parsed_id.kind != kind:
                raise ValueError("structural snapshot: entry id kind must match entry kind")
            if not isinstance(relations_raw, list):
                raise ValueError("structural snapshot: entry relations must be a list")
            if uid in idx.by_uid:
                raise ValueError(f"structural snapshot: duplicate uid {uid!r}")
            deprecated_ids = _validated_deprecated_ids(raw["deprecated_ids"], entry_id)
            relations = []
            for raw_relation in relations_raw:
                if not isinstance(raw_relation, dict):
                    raise ValueError("structural snapshot: relation row must be a dict")
                _validate_snapshot_directed(raw_relation, "relation row")
                _validate_snapshot_weight(raw_relation, "relation row")
                relation_data = dict(raw_relation)
                relation_data["attrs"] = deepcopy(relation_data.get("attrs", {}))
                try:
                    relations.append(Relation(**relation_data))
                except (PydanticValidationError, TypeError) as exc:
                    raise ValueError("structural snapshot: invalid relation row") from exc
            structural_refs = _validated_structural_refs(structural_refs_raw)
            out_refs = _relation_out_refs(relations) + structural_refs
            entry = IndexEntry(
                uid=uid,
                id=entry_id,
                kind=kind,
                deprecated_ids=frozenset(deprecated_ids),
                out_refs=out_refs,
            )
            for claim in (entry.id, *entry.deprecated_ids):
                owner = idx.resolve_uid(claim)
                if owner is not None:
                    raise ValueError(
                        f"structural snapshot: identity claim {claim!r} already in use by uid {owner!r}"
                    )
            idx.by_uid[uid] = entry
            idx.id_to_uid[entry.id] = uid
            for dep in entry.deprecated_ids:
                idx.deprecated_to_uid[dep] = uid
            for oref in out_refs:
                idx.in_refs.setdefault(oref.ref, []).append(InRef(source_uid=uid, out_ref=oref))
        return idx

    def remove(self, uid: str) -> None:
        self._drop(uid)

    def _drop(self, uid: str) -> None:
        entry = self.by_uid.pop(uid, None)
        if entry is None:
            return
        if self.id_to_uid.get(entry.id) == uid:
            del self.id_to_uid[entry.id]
        for dep in entry.deprecated_ids:
            if self.deprecated_to_uid.get(dep) == uid:
                del self.deprecated_to_uid[dep]
        for ref, rows in list(self.in_refs.items()):
            kept = [r for r in rows if r.source_uid != uid]
            if kept:
                self.in_refs[ref] = kept
            else:
                del self.in_refs[ref]

    def _refs_for_uid(self, uid: str) -> list[str]:
        entry = self.by_uid[uid]
        return [entry.id, *sorted(entry.deprecated_ids)]

    def _resolve_edge(self, rel: Relation) -> ResolvedEdge:
        return ResolvedEdge(
            relation=rel,
            source_uid=self.resolve_uid(rel.source),
            target_uid=self.resolve_uid(rel.target),
        )

    def _relations_by_role(self, uid: str, role: Role) -> list[ResolvedEdge]:
        seen: set[int] = set()
        edges: list[ResolvedEdge] = []
        for ref in self._refs_for_uid(uid):
            for inref in self.in_refs.get(ref, []):
                oref = inref.out_ref
                if oref.role != role or oref.relation is None:
                    continue
                if id(oref.relation) in seen:
                    continue
                seen.add(id(oref.relation))
                edges.append(self._resolve_edge(oref.relation))
        return edges

    def outbound_edges(self, uid: str) -> list[ResolvedEdge]:
        return self._relations_by_role(uid, "relation_source")

    def inbound_edges(self, uid: str) -> list[ResolvedEdge]:
        return self._relations_by_role(uid, "relation_target")

    def dangling_edges(self) -> list[ResolvedEdge]:
        seen: set[int] = set()
        edges: list[ResolvedEdge] = []
        for entry in self.by_uid.values():
            for oref in entry.out_refs:
                if oref.role != "relation_target" or oref.relation is None:
                    continue
                if id(oref.relation) in seen:
                    continue
                if self.resolve_uid(oref.ref) is None:
                    seen.add(id(oref.relation))
                    edges.append(self._resolve_edge(oref.relation))
        return edges

    def members_of(self, uid: str) -> set[str]:
        """Uids of this node's resolvable direct members. Dangling member refs are skipped
        (check reports them); duplicate entries and live+deprecated refs dedupe by uid."""
        entry = self.by_uid[uid]
        members: set[str] = set()
        for oref in entry.out_refs:
            if oref.role != "membership_member":
                continue
            member_uid = self.resolve_uid(oref.ref)
            if member_uid is not None:
                members.add(member_uid)
        return members

    def containers_of(self, uid: str) -> set[str]:
        """Uids of the nodes whose membership facet lists any of this node's identity claims
        (live id or deprecated ids — the same attribution rule as _relations_by_role)."""
        containers: set[str] = set()
        for ref in self._refs_for_uid(uid):
            for inref in self.in_refs.get(ref, []):
                if inref.out_ref.role != "membership_member":
                    continue
                containers.add(inref.source_uid)
        return containers

    def dangling_members(self) -> list[tuple[str, str]]:
        """Every unresolved membership ref, deduped by (container uid, ref)."""
        out: list[tuple[str, str]] = []
        seen: set[tuple[str, str]] = set()
        for entry in self.by_uid.values():
            for oref in entry.out_refs:
                if oref.role != "membership_member":
                    continue
                if self.resolve_uid(oref.ref) is not None:
                    continue
                key = (entry.uid, oref.ref)
                if key in seen:
                    continue
                seen.add(key)
                out.append(key)
        return out

File: ts/src/structural-index.ts
import { CollisionError, IdError, RefError } from "./errors.js";
import { NodeId } from "./ids.js";
import type { Node } from "./node.js";
import { pathCollisionKey, pathForNodeId } from "./paths.js";
import { type Relation, RelationSchema } from "./relations.js";
import { EDGES, KEYS, MEMBERSHIP, ORDER } from "./shapes.js";

export type Role =
  | "relation_source"
  | "relation_target"
  | "membership_member"
  | "edges_source"
  | "edges_target"
  | "order_member"
  | "keys_value";

const STRUCTURAL_ENTRY_KEYS = ["uid", "id", "kind", "deprecatedIds", "relations", "structuralRefs"] as const;
const STRUCTURAL_REF_ROLES = new Set<Role>([
  "membership_member",
  "edges_source",
  "edges_target",
  "order_member",
  "keys_value",
]);

export interface OutRef {
  ref: string;
  role: Role;
  relation?: Relation; // present iff role starts with "relation_"
}

export interface InRef {
  sourceUid: string;
  outRef: OutRef;
}

export interface IndexEntry {
  uid: string;
  id: string;
  kind: string;
  deprecatedIds: ReadonlySet<string>;
  outRefs: OutRef[];
}

export interface ResolvedEdge {
  relation: Relation;
  sourceUid: string | null;
  targetUid: string | null; // null when the endpoint ref is dangling
}

function relationOutRefs(relations: Relation[]): OutRef[] {
  const refs: OutRef[] = [];
  for (const rel of relations) {
    refs.push({ ref: rel.source, role: "relation_source", relation: rel });
    refs.push({ ref: rel.target, role: "relation_target", relation: rel });
  }
  return refs;
}

// Refs from the built-in structural facets. Read directly from `node.facets`
// (registry-independent); they populate `inRefs` for rename + dangling integrity but
// are never relation-graph edges (their `relation` stays undefined).
function structuralOutRefs(node: Node): OutRef[] {
  const refs: OutRef[] = [];
  const mem = node.facets[MEMBERSHIP];
  if (mem !== null && typeof mem === "object") {
    const members = (mem as Record<string, unknown>).members;
    if (Array.isArray(members)) {
      for (const m of members) if (typeof m === "string") refs.push({ ref: m, role: "membership_member" });
    }
  }
  const eg = node.facets[EDGES];
  if (eg !== null && typeof eg === "object") {
    const edges = (eg as Record<string, unknown>).edges;
    if (Array.isArray(edges)) {
      for (const edge of edges) {
        if (edge !== null && typeof edge === "object") {
          const e = edge as Record<string, unknown>;
          if (typeof e.source === "string") refs.push({ ref: e.source, role: "edges_source" });
          if (typeof e.target === "string") refs.push({ ref: e.target, role: "edges_target" });
        }
      }
    }
  }
  const od = node.facets[ORDER];
  if (od !== null && typeof od === "object") {
    const order = (od as Record<string, unknown>).order;
    if (Array.isArray(order)) {
      for (const m of order) if (typeof m === "string") refs.push({ ref: m, role: "order_member" });
    }
  }
  const ky = node.facets[KEYS];
  if (ky !== null && typeof ky === "object") {
    const keys = (ky as Record<string, unknown>).keys;
    if (keys !== null && typeof keys === "object") {
      for (const v of Object.values(keys as Record<string, unknown>)) {
        if (typeof v === "string") refs.push({ ref: v, role: "keys_value" });
      }
    }
  }
  return refs;
}

function extractOutRefs(node: Node): OutRef[] {
  return [...relationOutRefs(node.relations), ...structuralOutRefs(node)];
}

function validatedDeprecatedIds(raw: unknown, entryId: string): string[] {
  if (!Array.isArray(raw)) throw new Error("structural snapshot: deprecatedIds must be an array of strings");
  const out: string[] = [];
  const seen = new Set<string>();
  for (const dep of raw) {
    if (typeof dep !== "string") throw new Error("structural snapshot: deprecatedIds must be an array of strings");
    if (dep === entryId)
      throw new Error(
        `structural snapshot: identity claim ${JSON.stringify(dep)} is both live and deprecated in one entry`,
      );
    if (seen.has(dep))
      throw new Error(`structural snapshot: duplicate deprecated identity claim ${JSON.stringify(dep)} in one entry`);
    seen.add(dep);
    out.push(dep);
  }
  return out;
}

function validateSnapshotWeight(raw: Record<string, unknown>, label: string): void {
  if (!("weight" in raw)) return;
  const weight = raw.weight;
  if (typeof weight !== "number" || !Number.isFinite(weight)) {
    throw new Error(`structural snapshot: ${label} weight must be a finite number`);
  }
}

function validateSnapshotDirected(raw: Record<string, unknown>, label: string): void {
  if ("directed" in raw && typeof raw.directed !== "boolean") {
    throw new Error(`structural snapshot: ${label} directed must be a boolean`);
  }
}

function validatedStructuralRefs(raw: unknown): OutRef[] {
  if (!Array.isArray(raw)) throw new Error("structural snapshot: structuralRefs must be an array");
  const out: OutRef[] = [];
  for (const item of raw) {
    if (item === null || typeof item !== "object")
      throw new Error("structural snapshot: structuralRef must be an object");
    const { ref, role } = item as Record<string, unknown>;
    if (typeof ref !== "string") throw new Error("structural snapshot: structuralRef ref must be a string");
    if (typeof role !== "string" || !STRUCTURAL_REF_ROLES.has(role as Role))
      throw new Error("structural snapshot: structuralRef role is invalid");
    out.push({ ref, role: role as Role });
  }
  return out;
}

/** In-memory structural index. Pure data; no file I/O. */
export class Index {
  byUid = new Map<string, IndexEntry>();
  idToUid = new Map<string, string>();
  deprecatedToUid = new Map<string, string>();
  inRefs = new Map<string, InRef[]>();
  // Derived: collision key (folded mapped path) -> live uids claiming it. Never
  // serialized; rebuilt from entries on restore.
  private pathUids = new Map<string, Set<string>>();

  static build(nodes: Iterable<Node>): Index {
    const idx = new Index();
    for (const node of nodes) {
      if (idx.byUid.has(node.uid)) {
        throw new CollisionError(`duplicate uid ${JSON.stringify(node.uid)} in corpus`);
      }
      // Construction checks identity only: a path-collided corpus is legal on the
      // volume that holds it and is reported by check(), not refused here.
      idx.assertIdentityClaims(node);
      idx.upsert(node);
    }
    return idx;
  }

  resolveUid(ref: string): string | null {
    return this.idToUid.get(ref) ?? this.deprecatedToUid.get(ref) ?? null;
  }

  // The mutation gate: identity claims, then mapped-path admission. `upsert` is
  // mechanical and never raises; `Corpus.add` calls this before `upsert`. Construction
  // calls `assertIdentityClaims` alone; `Corpus.rename` calls `assertPathAvailable`
  // alone (rename changes a uid's live id, which the identity check would reject).
  assertAddable(node: Node): void {
    this.assertIdentityClaims(node);
    this.assertPathAvailable(node.uid, node.id);
  }

  assertIdentityClaims(node: Node): void {
    const existing = this.byUid.get(node.uid);
    if (existing !== undefined && existing.id !== node.id) {
      throw new CollisionError(
        `uid ${JSON.stringify(node.uid)} already belongs to live id ${JSON.stringify(existing.id)}; use rename()`,
      );
    }
    for (const claim of [node.id, ...node.deprecatedIds]) {
      const owner = this.resolveUid(claim);
      if (owner !== null && owner !== node.uid) {
        throw new CollisionError(
          `identity claim ${JSON.stringify(claim)} already in use by uid ${JSON.stringify(owner)}`,
        );
      }
    }
  }

  /** Mutation admission: refuse a mapped path whose collision key another live uid
   * already claims. A uid re-claiming its own exact mapped path changes nothing. */
  assertPathAvailable(candidateUid: string, candidateId: string): void {
    const candidatePath = pathForNodeId(candidateId);
    const existing = this.byUid.get(candidateUid);
    if (existing !== undefined && pathForNodeId(existing.id) === candidatePath) return;
    const key = pathCollisionKey(candidateId);
    if (this.pathUids.has(key)) {
      throw new CollisionError(`mapped path for ${JSON.stringify(candidateId)} collides at ${JSON.stringify(key)}`);
    }
  }

  /** Every [live id, collision key] in a bucket claimed by more than one uid. */
  pathCollisions(): Array<[string, string]> {
    const rows: Array<[string, string]> = [];
    for (const [key, uids] of this.pathUids) {
      if (uids.size < 2) continue;
      for (const uid of uids) rows.push([(this.byUid.get(uid) as IndexEntry).id, key]);
    }
    return rows;
  }

  upsert(node: Node): void {
    if (this.byUid.has(node.uid)) this.drop(node.uid);
    const entry: IndexEntry = {
      uid: node.uid,
      id: node.id,
      kind: node.kind,
      deprecatedIds: new Set(node.deprecatedIds),
      outRefs: extractOutRefs(node),
    };
    this.byUid.set(node.uid, entry);
    this.claimPath(entry);
    this.idToUid.set(node.id, node.uid);
    for (const dep of node.deprecatedIds) this.deprecatedToUid.set(dep, node.uid);
    for (const oref of entry.outRefs) {
      const rows = this.inRefs.get(oref.ref) ?? [];
      rows.push({ sourceUid: node.uid, outRef: oref });
      this.inRefs.set(oref.ref, rows);
    }
  }

  toDict(): {
    entries: Array<{
      uid: string;
      id: string;
      kind: string;
      deprecatedIds: string[];
      relations: Array<Record<string, unknown>>;
      structuralRefs: Array<{ ref: string; role: Role }>;
    }>;
  } {
    const entries = [];
    for (const entry of this.byUid.values()) {
      const relations: Array<Record<string, unknown>> = [];
      for (const o of entry.outRefs) {
        if (o.role !== "relation_source" || o.relation === undefined) continue;
        const rel = o.relation;
        const row: Record<string, unknown> = {
          source: rel.source,
          predicate: rel.predicate,
          target: rel.target,
          directed: rel.directed,
          attrs: structuredClone(rel.attrs),
        };
        if (rel.weight !== null) row.weight = rel.weight;
        relations.push(row);
      }
      entries.push({
        uid: entry.uid,
        id: entry.id,
        kind: entry.kind,
        deprecatedIds: [...entry.deprecatedIds].sort(),
        relations,
        structuralRefs: entry.outRefs
          .filter((o) => !o.role.startsWith("relation_"))
          .map((o) => ({ ref: o.ref, role: o.role })),
      });
    }
    return { entries };
  }

  static fromDict(d: unknown): Index {
    if (typeof d !== "object" || d === null) throw new Error("structural snapshot: document must be an object");
    const raw = d as Record<string, unknown>;
    if (!("entries" in raw)) throw new Error("structural snapshot: missing entries");
    if (!Array.isArray(raw.entries)) throw new Error("structural snapshot: entries must be an array");
    const idx = new Index();
    for (const rawEntry of raw.entries) {
      if (typeof rawEntry !== "object" || rawEntry === null)
        throw new Error("structural snapshot: entry must be an object");
      const e = rawEntry as Record<string, unknown>;
      for (const key of STRUCTURAL_ENTRY_KEYS) {
        if (!(key in e)) throw new Error(`structural snapshot: entry missing ${key}`);
      }
      const { uid, id: entryId, kind } = e;
      if (typeof uid !== "string" || uid.length === 0) {
        throw new Error("structural snapshot: entry uid must be a non-empty string");
      }
      if (typeof entryId !== "string") throw new Error("structural snapshot: entry id must be a string");
      if (typeof kind !== "string") throw new Error("structural snapshot: entry kind must be a string");
      let parsed: NodeId;
      try {
        parsed = NodeId.parse(entryId);
      } catch (err) {
        if (err instanceof IdError) throw new Error("structural snapshot: entry id must be a valid node id");
        throw err;
      }
      if (parsed.kind !== kind) throw new Error("structural snapshot: entry id kind must match entry kind");
      if (!Array.isArray(e.relations)) throw new Error("structural snapshot: entry relations must be an array");
      if (idx.byUid.has(uid)) throw new Error(`structural snapshot: duplicate uid ${JSON.stringify(uid)}`);
      const deprecatedIds = validatedDeprecatedIds(e.deprecatedIds, entryId);
      const relations: Relation[] = [];
      for (const rawRel of e.relations) {
        if (typeof rawRel !== "object" || rawRel === null)
          throw new Error("structural snapshot: relation row must be an object");
        const rr = rawRel as Record<string, unknown>;
        validateSnapshotDirected(rr, "relation row");
        validateSnapshotWeight(rr, "relation row");
        const relData = { ...rr, attrs: structuredClone(rr.attrs ?? {}) };
        const result = RelationSchema.safeParse(relData);
        if (!result.success) throw new Error("structural snapshot: invalid relation row");
        relations.push(result.data);
      }
      const structuralRefs = validatedStructuralRefs(e.structuralRefs);
      const outRefs = [...relationOutRefs(relations), ...structuralRefs];
      const entry: IndexEntry = {
        uid,
        id: entryId,
        kind,
        deprecatedIds: new Set(deprecatedIds),
        outRefs,
      };
      for (const claim of [entry.id, ...entry.deprecatedIds]) {
        const owner = idx.resolveUid(claim);
        if (owner !== null) {
          throw new Error(
            `structural snapshot: identity claim ${JSON.stringify(claim)} already in use by uid ${JSON.stringify(owner)}`,
          );
        }
      }
      idx.byUid.set(uid, entry);
      idx.claimPath(entry);
      idx.idToUid.set(entry.id, uid);
      for (const dep of entry.deprecatedIds) idx.deprecatedToUid.set(dep, uid);
      for (const oref of outRefs) {
        const rows = idx.inRefs.get(oref.ref) ?? [];
        rows.push({ sourceUid: uid, outRef: oref });
        idx.inRefs.set(oref.ref, rows);
      }
    }
    return idx;
  }

  remove(uid: string): void {
    this.drop(uid);
  }

  private claimPath(entry: IndexEntry): void {
    const key = pathCollisionKey(entry.id);
    const bucket = this.pathUids.get(key) ?? new Set<string>();
    bucket.add(entry.uid);
    this.pathUids.set(key, bucket);
  }

  private drop(uid: string): void {
    const entry = this.byUid.get(uid);
    if (entry === undefined) return;
    this.byUid.delete(uid);
    const key = pathCollisionKey(entry.id);
    const bucket = this.pathUids.get(key) as Set<string>;
    bucket.delete(uid);
    if (bucket.size === 0) this.pathUids.delete(key);
    if (this.idToUid.get(entry.id) === uid) this.idToUid.delete(entry.id);
    for (const dep of entry.deprecatedIds) {
      if (this.deprecatedToUid.get(dep) === uid) this.deprecatedToUid.delete(dep);
    }
    // Drop only the refs THIS node contributed as a source. Inbound refs that other
    // (surviving) nodes contributed pointing at this node's ids must persist — they
    // are still on disk and become dangling.
    for (const [ref, rows] of [...this.inRefs.entries()]) {
      const kept = rows.filter((r) => r.sourceUid !== uid);
      if (kept.length > 0) this.inRefs.set(ref, kept);
      else this.inRefs.delete(ref);
    }
  }

  private refsForUid(uid: string): string[] {
    const entry = this.byUid.get(uid);
    if (entry === undefined) throw new RefError(`uid ${JSON.stringify(uid)} not in index`);
    return [entry.id, ...[...entry.deprecatedIds].sort()];
  }

  private resolveEdge(rel: Relation): ResolvedEdge {
    return { relation: rel, sourceUid: this.resolveUid(rel.source), targetUid: this.resolveUid(rel.target) };
  }

  // Public graph queries are defined over distinct `Relation` OBJECTS (reference identity —
  // the TS analog of Python's `id(relation)` dedup), relations-only. A relation never
  // appears twice, and a relation whose source is a non-container node still attributes
  // correctly because we key on `relation_source` / `relation_target` roles.
  private relationsByRole(uid: string, role: Role): ResolvedEdge[] {
    const seen = new Set<Relation>();
    const edges: ResolvedEdge[] = [];
    for (const ref of this.refsForUid(uid)) {
      for (const inref of this.inRefs.get(ref) ?? []) {
        const oref = inref.outRef;
        if (oref.role !== role || oref.relation === undefined) continue;
        if (seen.has(oref.relation)) continue;
        seen.add(oref.relation);
        edges.push(this.resolveEdge(oref.relation));
      }
    }
    return edges;
  }

  outboundEdges(uid: string): ResolvedEdge[] {
    return this.relationsByRole(uid, "relation_source");
  }

  inboundEdges(uid: string): ResolvedEdge[] {
    return this.relationsByRole(uid, "relation_target");
  }

  danglingEdges(): ResolvedEdge[] {
    const seen = new Set<Relation>();
    const edges: ResolvedEdge[] = [];
    for (const entry of this.byUid.values()) {
      for (const oref of entry.outRefs) {
        // Only the target side can dangle: a relation is dangling when its TARGET ref resolves
        // to no live uid. (Source-side refs are the node's own outbound edges, always resolvable.)
        if (oref.role !== "relation_target" || oref.relation === undefined) continue;
        if (seen.has(oref.relation)) continue;
        if (this.resolveUid(oref.ref) === null) {
          seen.add(oref.relation);
          edges.push(this.resolveEdge(oref.relation));
        }
      }
    }
    return edges;
  }

  /** Uids of this node's resolvable direct members. Dangling member refs are skipped
   * (check reports them); duplicate entries and live+deprecated refs dedupe by uid. */
  membersOf(uid: string): Set<string> {
    const entry = this.byUid.get(uid);
    if (entry === undefined) throw new RefError(`uid ${JSON.stringify(uid)} not in index`);
    const members = new Set<string>();
    for (const oref of entry.outRefs) {
      if (oref.role !== "membership_member") continue;
      const memberUid = this.resolveUid(oref.ref);
      if (memberUid !== null) members.add(memberUid);
    }
    return members;
  }

  /** Uids of the nodes whose membership facet lists any of this node's identity claims
   * (live id or deprecated ids — the same attribution rule as relationsByRole). */
  containersOf(uid: string): Set<string> {
    const containers = new Set<string>();
    for (const ref of this.refsForUid(uid)) {
      for (const inref of this.inRefs.get(ref) ?? []) {
        if (inref.outRef.role !== "membership_member") continue;
        containers.add(inref.sourceUid);
      }
    }
    return containers;
  }

  /** Every unresolved membership ref, deduped by (container uid, ref). */
  danglingMembers(): Array<{ sourceUid: string; ref: string }> {
    const out: Array<{ sourceUid: string; ref: string }> = [];
    const seen = new Set<string>();
    for (const entry of this.byUid.values()) {
      for (const oref of entry.outRefs) {
        if (oref.role !== "membership_member") continue;
        if (this.resolveUid(oref.ref) !== null) continue;
        const key = `${entry.uid}\u0000${oref.ref}`;
        if (seen.has(key)) continue;
        seen.add(key);
        out.push({ sourceUid: entry.uid, ref: oref.ref });
      }
    }
    return out;
  }
}
